### PR TITLE
Issue 43: Validation of document definitions with Joi

### PR DESCRIPTION
**Pull request by OldSneerJaw**
Introduces a complete schema for document definitions, built with [Joi](https://github.com/hapijs/joi). To be executed by the `validate-document-definitions` script, which makes use of the new `src/validation/document-definitions-validator` module.

This pull request is the first step to address issue #43. Subsequent PRs will remove the previous work-in-progress validation code and statically embed the Joi package in the repo's `lib` directory to preserve synctos' status of having no external runtime dependencies.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,7 +31,15 @@ As a convenience, otto - and, by extension, Sync Gateway - does support the [Und
 Once synctos is installed, you can run it from your project's directory as follows:
 
 ```
-node_modules/synctos/make-sync-function /path/to/my-document-definitions.js /path/to/my-generated-sync-function.js
+node_modules/.bin/synctos /path/to/my-document-definitions.js /path/to/my-generated-sync-function.js
```
Comment (OldSneerJaw) [inline README.md]: This is made possible by the inclusion of a [bin](https://docs.npmjs.com/files/package.json#bin) entry in `package.json`.

File: README.md [hunk 2]
```diff
@@ -40,6 +48,24 @@ Generated sync functions are compatible with all Sync Gateway 1.x versions.
 
 **NOTE**: Due to a [known issue](https://github.com/couchbase/sync_gateway/issues/1866) in Sync Gateway versions up to and including 1.2.1, when specifying a bucket/database's sync function in a configuration file as a multi-line string, you will have to be sure to escape any literal backslash characters in the sync function body. For example, if your sync function contains a regular expression like `new RegExp('\\w+')`, you will have to escape the backslashes when inserting the sync function into the configuration file so that it becomes `new RegExp('\\\\w+')`. The issue has been resolved in Sync Gateway version 1.3.0 and later.
 
+### Validating
+
+To validate that your document definitions file is structured correctly and does not contain any obvious semantic violations, execute the built in validation script as follows:
+
+```
+node_modules/.bin/synctos-validate /path/to/my-document-definitions.js
```
Comment (OldSneerJaw) [inline README.md]: This is made possible by the inclusion of a [bin](https://docs.npmjs.com/files/package.json#bin) entry in `package.json`.

File: etc/prepare-tests.sh [hunk 3]
```diff
@@ -10,12 +10,13 @@ mkdir -p build/test-reports/
 # Set up JSHint configuration for the generated sync functions
 cp "etc/jshintrc-sync-function-template.json" "$outputDir/.jshintrc"
 
-echo "Linting modules and specs with JSHint\n"
-node_modules/jshint/bin/jshint src/*.js test/*.js
+echo "Linting modules and specs with JSHint...\n"
+node_modules/jshint/bin/jshint src test/*.js
```
Comment (OldSneerJaw) [inline etc/prepare-tests.sh]: Specifying the `src` directory as the target directly ensures that JSHint recurses into subdirectories.

File: package.json [hunk 4]
```diff
@@ -2,7 +2,9 @@
   "name": "synctos",
   "version": "1.10.0",
   "main": "src/index.js",
-  "dependencies": {},
+  "dependencies": {
+    "joi": "^13.1.1"
```
Comment (OldSneerJaw) [inline package.json]: As I mentioned in the pull request description, I will remove this dependency in a subsequent pull request that will embed it as a static library in the `lib` directory instead.

File: package.json [hunk 5]
```diff
@@ -11,12 +13,16 @@
     "nyc": "^11.4.1"
   },
   "scripts": {
-    "test": "etc/prepare-tests.sh && nyc mocha **/*.spec.js",
-    "test-report": "etc/prepare-tests.sh && nyc mocha -R xunit **/*.spec.js > build/test-reports/synctos.xml"
+    "test": "etc/prepare-tests.sh && nyc mocha \"**/*.spec.js\"",
+    "test-report": "etc/prepare-tests.sh && nyc mocha -R xunit \"**/*.spec.js\" > build/test-reports/synctos.xml"
```
Comment (OldSneerJaw) [inline package.json]: This change allows Mocha to recurse into subdirectories (e.g. the new `src/validation` directory).

File: package.json [hunk 6]
```diff
@@ -11,12 +13,16 @@
...
   },
   "license": "MIT",
   "repository": {
     "type": "git",
     "url": "https://github.com/Kashoo/synctos"
+  },
+  "bin": {
+    "synctos": "./make-sync-function",
+    "synctos-validate": "./validate-document-definitions"
```
Comment (OldSneerJaw) [inline package.json]: These entries ensure that a project that depends on synctos will automatically install scripts with the specified names into the dependent project's `node_modules/.bin` directory.

File: src/validation/document-definitions-validator.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+var environmentShellMaker = require('../document-definitions-shell-maker');
```
Comment (OldSneerJaw) [inline src/validation/document-definitions-validator.js]: This is intended to take the place of the work-in-progress `src/document-definitions-validator` module (notice the difference in file paths). The WIP module will be removed in a different pull request.

File: package.json (hunk 6 above)
Comment (dkichler) [reply package.json]: slick az

File: src/validation/property-validator-schema.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+      minimumValue: dynamicConstraintSchema(joi.number()),
+      minimumValueExclusive: dynamicConstraintSchema(joi.number()),
+      maximumValue: maximumValueInclusiveNumberConstraintSchema(joi.number()),
+      maximumValueExclusive: maximumValueExclusiveNumberConstraintSchema(joi.number())
+    },
+    boolean: { },
+    datetime: {
+      minimumValue: dynamicConstraintSchema(datetimeSchema),
+      minimumValueExclusive: dynamicConstraintSchema(datetimeSchema),
+      maximumValue: dynamicConstraintSchema(datetimeSchema),
```
[the remark this answers is not in the document]
Comment (OldSneerJaw) [reply src/validation/property-validator-schema.js]: Joi only exposes range constraints (e.g. `greater`, `less`, etc.) for numeric types, so other types have to do without.

File: src/validation/property-validator-schema.js [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+      minimumValue: dynamicConstraintSchema(dateOnlySchema),
+      minimumValueExclusive: dynamicConstraintSchema(dateOnlySchema),
+      maximumValue: dynamicConstraintSchema(dateOnlySchema),
+      maximumValueExclusive: dynamicConstraintSchema(dateOnlySchema)
+    },
+    enum: {
+      predefinedValues: dynamicConstraintSchema(joi.array().required().min(1).items([ integerSchema, joi.string() ]))
+    },
+    uuid: {
+      minimumValue: dynamicConstraintSchema(uuidSchema),
```
[the remark this answers is not in the document]
Comment (OldSneerJaw) [reply src/validation/property-validator-schema.js]: I figure it doesn't hurt to allow someone to set a range of valid UUIDs. I'll update the README to include them.

File: src/validation/property-validator-schema.js [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+// Defined as a function rather than a plain object because it contains lazy references that result in recursive
+// references between the complex types (e.g. "array", "object", "hashtable") and the main "propertyValidators" schema
+function typeSpecificConstraintSchemas() {
+  return {
+    string: {
+      mustNotBeEmpty: dynamicConstraintSchema(joi.boolean()),
+      regexPattern: dynamicConstraintSchema(regexSchema),
+      minimumLength: dynamicConstraintSchema(integerSchema.min(0)),
+      maximumLength: maximumSizeConstraintSchema('minimumLength'),
+      minimumValue: dynamicConstraintSchema(joi.string()),
```
[the remark this answers is not in the document]
Comment (OldSneerJaw) [reply src/validation/property-validator-schema.js]: Technically the constraints are allowed on strings (and any other data type). I figured it doesn't hurt to allow them in order to define a range of valid string values. For example `minimumValue: 'A', maximumValueExclusive: 'G' would allow only strings that start with the letters "A" through "F". I'll document them in the README.

File: src/validation/property-validator-schema.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (OldSneerJaw) [reply src/validation/property-validator-schema.js]: Correction: Joi's [date](https://github.com/hapijs/joi/blob/v13.1.1/API.md#date---inherits-from-any) type _does_ support `min` and `max` values (but not `greater` or `less`). I will investigate if it makes sense to use them.

[the remark this answers is not in the document]
Comment (OldSneerJaw) [reply src/validation/property-validator-schema.js]: Er... now I remember why I chose to use a string with a regex rather than [`joi.date()`](https://github.com/hapijs/joi/blob/v13.1.1/API.md#date---inherits-from-any) for the `date` and `datetime` data types: when Joi is running, it is doing so in a Node.js process, which supports a different subset of the ISO 8601 date-time string format than Sync Gateway's JavaScript interpreter does. It ends up allowing date strings that cannot be parsed when the generated sync function is used in Sync Gateway. I'll put a comment in the code to explain this.

**Comment by OldSneerJaw** [comment]
@dkichler: I have addressed your comments. Ready for another review.

File: src/validation/property-validator-schema.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (dkichler) [reply src/validation/property-validator-schema.js]: 👍

File: src/validation/property-validator-schema.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dkichler) [reply src/validation/property-validator-schema.js]: 👍

File: src/validation/property-validator-schema.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dkichler) [reply src/validation/property-validator-schema.js]: 👍

**Comment by dkichler** [comment]
just re-read the description denoting this as the first of several toward #43 😞, maybe I'll just hack it to test our doc defns right away 🙏
